torch/distributed/internal_rpc_utils.py

```diff
@@ -5,28 +5,54 @@
 import io
 import pickle
 import traceback
+import threading
+import sys
 
+import torch
+
+_thread_local_tensor_tables = threading.local()
+_thread_local_tensor_tables.recv_tables = []
+_thread_local_tensor_tables.send_tables = []
+
+
+def _tensor_receiver(tensor_index):
+    global _thread_local_tensor_tables
+    return _thread_local_tensor_tables.recv_tables[tensor_index]
+
+def tensor_reducer(obj):
+    global _thread_local_tensor_tables
+    _thread_local_tensor_tables.send_tables.append(obj)
+    tensor_index = len(_thread_local_tensor_tables.send_tables) - 1
+    return (_tensor_receiver, (tensor_index, ))
+
+if sys.version_info >= (3, 0):
+    _dispatch_table = copyreg.dispatch_table.copy()
+    _dispatch_table[torch.Tensor] = tensor_reducer
 
 def serialize(obj):
     f = io.BytesIO()
     p = pickle.Pickler(f)
-    p.dispatch_table = copyreg.dispatch_table.copy()
+    p.dispatch_table = _dispatch_table
+    global _thread_local_tensor_tables
+    _thread_local_tensor_tables.send_tables = []
```

Comment:
It's not immediately clear why this is needed. I see it's related to the reducer in the dispatch table. Would it be possible to turn the dispatch table into a first class object and associate some methods with it? E.g. `_dispatch_table.clear_send_tables()` or something? Then we can keep all the logic related to pickling/unpickling in one place vs. mucking with globals.

Reply:
@pietern in order to achieve that we'd need a per-serialization dispatch table (or at least a thread-local dispatch table). I don't know what's the relative cost of copying the dispatch table, but I'd rather have a global dispatch table and a thread-local pickling context object instead.

===

On a separate note, I believe that there is a corner case not covered here: it is possible for a custom reducer to call an RPC. In that case you'll have nested calls to serialize on the same thread. Example:

```
class MyCustomClass:
    def __getstate__(self):
        dist.rpc(my_random_remote_function_called_inside_of_serialize, ..., to='w1')
        return (...)

m = MyCustomClass()
rpc(my_function_taking_a_custom_class_instance, args=(m, ), 'w1')
```

In order to cover this corner case you'll need to save a possible current _thread_local_tensor_tables.send_tables and restore it after the call. Even if we go with alternative solutions such as suggested by Pieter we'll have to deal with this case.